gymnasium/gymstatic
# [Release]: Jobs Feed Improvements

Please see https://github.com/gymnasium/gymcms/pull/860 for the initial PR.

---

### The Current Problem(s)
A. The jobs feed seems to run into issues with reliably fetching jobs from the Aquent endpoint, and it seems to be because of CORS errors. This is surprising and frustrating because the our sites are already on the allow list for cross-origin requests. As a result of these errors, the job feed has been consistently using the fallback feed, which is only as recent as the last gymcms build (for each corresponding server - dev, staging, production).
B. For some reason, the date sorting provided by the dotCMS feed is wonky at best. We have the potential to sort by `posted date` or `mod date` and the feed doesn't do it properly.

### What this update does:
- Avoids CORS errors by using JSONP to reliably fetch jobs. It's kind of an old school method at this point, but it seems to work better than our previous approaches.
- Implements proper sorting by date - we compare the `posted date` and the `mod date` to see which is more recent.
- If the primary feed fails, it attempts to pull in the feed generated during the last gymcms (environment) build.
- shows an error message to the user if fetching the secondary fallback feed results in an error (ie, 200, 404, etc). The user is provided a link to the Aquent job feed.

### Testing:
- Visit https://courses.gymna.si/jobs in an incognito window (I'm using Chrome's inspector, so you'll have to adjust the tab names specified below for other browsers)
- open up inspector (right click anywhere on the page).
- Click on the "Application" tab, and look for "Session Storage". Open it up and look for a key named "jobs". Click on the key and look at the `date_modified` field - it should be roughly the time you opened up the page. Now, select the key again and delete it. Yes, delete it. We're going to have you load the jobs anew while looking at the console.
- click on the "Console" tab. In the upper left corner is an icon that looks like a circle with a diagonal line through it (like a stop icon). Click on it, and this will clear the console.
- Add ?debug=true to the URL, like so `https://courses.gymna.si/jobs?debug=true` and click enter. Make a note of the current time, by the way.
- You should see several lines of stuff with the prefix `job module`, and there should be a reference to `JSONP` in there.  Try messing around with the location dropdown, and you should see all of this debug data change to match your selection.
- Go back to the "Application" tab and look at the session storage again.  Check the date/time stamp and it should match when you added the debug option to the URL.

### Negative Outcome Tests
Remove the session storage object and make sure the console is open before visiting each of these pages.
- https://staging.thegymcms.com/tests/jobs/broken-feed
- https://staging.thegymcms.com/tests/jobs/broken-fallback (this result should display an error message to the end user).

---
See https://github.com/gymnasium/tracker/issues/322 for additional details.
